cs-excel/src/main/java/io/cloudslang/content/excel/services/ExcelServiceImpl.java

```diff
@@ -0,0 +1,360 @@
+import org.apache.poi.ss.usermodel.CellType;
+import org.apache.poi.ss.usermodel.DateUtil;
+import org.apache.poi.ss.usermodel.Row;
+import org.apache.poi.ss.usermodel.Sheet;
+import org.apache.poi.ss.usermodel.Workbook;
+import org.apache.poi.ss.usermodel.WorkbookFactory;
+
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+import static io.cloudslang.content.excel.utils.Constants.BAD_EXCEL_FILE_MSG;
+import static io.cloudslang.content.excel.utils.Constants.FORMAT_XLS;
+import static io.cloudslang.content.excel.utils.Constants.FORMAT_XLSM;
+import static io.cloudslang.content.excel.utils.Constants.FORMAT_XLSX;
+
+/**
+ * Created by danielmanciu on 18.02.2019.
+ */
+public class ExcelServiceImpl {
+
+    public static Workbook getWorkbook(final String fileName) throws Exception {
+        final String format = getFileFormat(fileName);
+        if (!isValidExcelFormat(format)) {
+            throw new InvalidFormatException(BAD_EXCEL_FILE_MSG);
+        }
+
+        FileInputStream input = null;
+        final Workbook excelDoc;
+
+        try {
+            input = new FileInputStream(fileName);
+            excelDoc = WorkbookFactory.create(input);
+            input.close();
+        } catch (IOException e) {
+            throw e;
+        } finally {
+            if (input != null) {
+                input.close();
+            }
+        }
+
+        return excelDoc;
+    }
+
+    public static boolean isValidExcelFormat(final String format) throws Exception {
+        if (StringUtils.isBlank(format)) {
+            throw new ExcelOperationException(BAD_EXCEL_FILE_MSG);
+        } else
+            return (format.equalsIgnoreCase(FORMAT_XLSX)
+                    || format.equalsIgnoreCase(FORMAT_XLS)
+                    || format.equalsIgnoreCase(FORMAT_XLSM));
+
+    }
+
+    public static void updateWorkbook(final Workbook workbook, final String fileName) throws IOException {
+        FileOutputStream output = null;
+        try {
+            output = new FileOutputStream(fileName);
+            workbook.write(output);
+            output.close();
+        } catch (IOException e) {
+            throw e;
+        } finally {
+            if (output != null) {
+                output.close();
+            }
+        }
+    }
+
+    public static String getFileFormat(final String excelFileName) {
+        final int index = excelFileName.lastIndexOf(".");
+        if (index > -1 && index < excelFileName.length()) {
+            return excelFileName.substring(index + 1);
+        }
+
+        return null;
+    }
+
+    public static Workbook getExcelDoc(final String excelFileName) {
+        Workbook excelDoc = null;
+        try {
+            excelDoc = getWorkbook(excelFileName);
+            if (excelDoc == null) {
+                throw new ExcelOperationException("Could not open " + excelFileName);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer

Exception [ExcelOperationException] is explicitly thrown on this line but is caught in the outer catch clause with generic exception on line [110].  Consider adding a catch clause of this specific exception in the outer try statement.

Reply:
Done